### PR TITLE
Add configuration for Hydra builds

### LINKED ISSUE #17: Setup CI
From #16 we'll have a script to typecheck all the generated files, and as mentioned in https://github.com/dhall-lang/dhall-kubernetes/pull/14#pullrequestreview-135229951 it would be nice to have CI check a set of end-to-end examples so that we don't inadvertently break users.

/cc @Gabriel439 should we use the Hydra instance as all the other repos, and if yes what's missing to make it work?

### PR DESCRIPTION
**Pull request by f-f**
Fix #17 

This adds a `release.nix` so that Hydra can start running our tests.

I needed a fresh version of Dhall (to run `dhall resolve` - I am really tempted about going for #7 now, as it would make this unnecessary, but we'll see) so I just pinned nixpkgs to the latest commit. This means that Hydra will have to compile some haskell packages (and Dhall itself), but I expect them to be cached in subsequent runs, right?

Also I integrated the `check-build.py` into the nix file, as it needed to use the nix-provided binaries (bash and dhall)

/cc @geigerzaehler

### CODE REVIEW

File: release.nix [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  nixpkgs = fetchNixpkgs {
+    rev = "c5f9cd4cde81bcef2c8882d359d9d30313bebeb3";
+
+    sha256 = "12w6g50hgk5nmijgs3rv7kq5mdybnkyv3cr2zm0q3ka9ax180dk6";
+
+    outputSha256 = "1nilhz6rhfg3ckp8yfmgy6v3q6spbyxfg0yn8rc2ydyn7119h4fn";
+  };
+
+  overlay = pkgsNew: pkgsOld: {
+    check-source = pkgsNew.writeText "check-source.py" ''
```
Comment (geigerzaehler) [inline release.nix]: Can’t we keep this as a file in the repo? It’s going to be a lot easier to maintain that way

Comment (f-f) [reply release.nix]: I'm not sure we can keep it in the repo, as to run the script with CI in nix, we need to patch the executables for `dhall` ([here](https://github.com/dhall-lang/dhall-kubernetes/pull/23/commits/597b56408e9711a71cfa73d08b03292bfef60f70#diff-6ab84d77bf6d73598ebba867195ce313R43)) and `bash` ([here](https://github.com/dhall-lang/dhall-kubernetes/pull/23/commits/597b56408e9711a71cfa73d08b03292bfef60f70#diff-6ab84d77bf6d73598ebba867195ce313R44)) (as nix keeps every version installed in a separate path, and in the job we deterministically specify the version of every dependency - as we pin the revision of the `nixpkgs` repo).

We could do it with `sed`, but I think it's fine (and equally easy to maintain) to keep it in the nix file, as:
- the users are not supposed to run the script, but it's just supposed to run in CI as a smoketest (will run on PR branches and `master`)
- if the users actually want to run it, it's just `nix-build release.nix` (we can add this to the readme I guess)
- my hope is that we could get rid of it entirely: if we go with #7 we won't need it, as we would just build the Dhall AST directly from the Haskell library (so it's correct by construction)

**Review by geigerzaehler — APPROVED** [review]

**Comment by Gabriella439** [comment]
Sorry for the delay on the Hydra status notification.  I ran into some issues on Hydra which I'm in the process of fixing

**Comment by Gabriella439** [comment]
@f-f: Alright, CI is fixed now.  This appears to be failing to due to locale errors on certain files:

```
<stdout>: hPutChar: invalid argument (invalid character)
```

This might be related to https://github.com/dhall-lang/dhall-haskell/issues/504

In the short term you might be able to fix the issue by explicitly setting the locale to UTF-8 in the test script.  I think the way to do this is to `export LANG="en_US.utf8"`, but I'm not sure.  Here is what my current locale is:

```
$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```

File: release.nix [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        "all-generated-files-compile"
+        { nativeBuildInputs = [
+            pkgsNew.python3
+            pkgsNew.bash
+            pkgsNew.dhall
+          ];
+        }
+        ''
+          cd ${./.}
+          python3 ${pkgsNew.check-source}
```
Comment (geigerzaehler) [inline release.nix]: Prepending `LC_ALL=en_US.UTF-8` seems to fix the locale issue.

**Comment by f-f** [comment]
@Gabriel439  @geigerzaehler 🤔 I tried setting those, but no luck. The internet suggests doing [stuff like this](https://github.com/tianon/docker-brew-debian/issues/45#issuecomment-325235517) but not sure it would work.

**Comment by Gabriella439** [comment]
The failures are consistent with a UTF-8 encoding issue since hand-checking a few results show that the failures exactly match the files that have unicode characters when pretty-printed.

I can reproduce the failure when I log into `hydra` and do `nix-store --realise` on the same derivation that CI built:

```
# nix-store --realise /nix/store/d47fzkvy0ccbh1j33jiljm6gnwi7c202-all-generated-files-compile.drv --keep-failed
these derivations will be built:
  /nix/store/d47fzkvy0ccbh1j33jiljm6gnwi7c202-all-generated-files-compile.drv
building '/nix/store/d47fzkvy0ccbh1j33jiljm6gnwi7c202-all-generated-files-compile.drv'...
Checking ./default/io.k8s.kubernetes.pkg.api.v1.Volume.dhall
Checking ./default/io.k8s.api.core.v1.ScaleIOPersistentVolumeSource.dhall
/nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
<stdout>: hPutChar: invalid argument (invalid character)
...
note: keeping build directory '/tmp/nix-build-all-generated-files-compile.drv-0'
builder for '/nix/store/d47fzkvy0ccbh1j33jiljm6gnwi7c202-all-generated-files-compile.drv' failed with exit code 1
error: build of '/nix/store/d47fzkvy0ccbh1j33jiljm6gnwi7c202-all-generated-files-compile.drv' failed

```

... but the first weird result is that when I try to reproduce the failure interactively using the same environment then everything works:

```
# cd /tmp/nix-build-all-generated-files-compile.drv-0
# source env-vars
# source $stdenv/setup
# genericBuild
Checking ./default/io.k8s.kubernetes.pkg.api.v1.Volume.dhall
Checking ./default/io.k8s.api.core.v1.ScaleIOPersistentVolumeSource.dhall
Checking ./default/io.k8s.api.authorization.v1beta1.ResourceAttributes.dhall
Checking ./default/io.k8s.api.storage.v1beta1.VolumeError.dhall
...
```

Another thing that is weird about this failure is that according to the error message it is failing to write a character to `stdout` but the `python` program has `stdout` configured to `DEVNULL`.

Also, the Hydra machine has the right locale set:

```
# locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=

# sudo -u nixbld1 locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

I'm still digging into this, but just wanted to document what I've learned so far.

**Comment by geigerzaehler** [comment]
I think I fixed the issue in #26. The problem seems to have been that the Hydra built hid the locale folder (which is usually something like `/usr/share/locale`). The fix was to just depend on `glibcLocales`. That package provides a nix setup hook that sets the `LOCALE_ARCHIVE` variable which includes the `en_US.UTF-8` locale.

**Pull request merged**